app/src/main/java/chat/rocket/android/helper/MessageParser.kt

```diff
@@ -0,0 +1,177 @@
+package chat.rocket.android.helper
+
+import android.app.Application
+import android.content.Context
+import android.graphics.Canvas
+import android.graphics.Color
+import android.graphics.Paint
+import android.graphics.Typeface
+import android.graphics.drawable.Drawable
+import android.text.Layout
+import android.text.Spannable
+import android.text.Spanned
+import android.text.TextPaint
+import android.text.style.*
+import android.view.View
+import chat.rocket.android.R
+import chat.rocket.android.chatroom.viewmodel.MessageViewModel
+import chat.rocket.core.model.url.Url
+import org.commonmark.node.BlockQuote
+import ru.noties.markwon.Markwon
+import ru.noties.markwon.SpannableBuilder
+import ru.noties.markwon.SpannableConfiguration
+import ru.noties.markwon.renderer.SpannableMarkdownVisitor
+import java.util.regex.Pattern
+import javax.inject.Inject
+
+
+class MessageParser @Inject constructor(val context: Application, private val configuration: SpannableConfiguration) {
+
+    private val parser = Markwon.createParser()
+    private val userPattern = Pattern.compile("(@[\\w.]+)",
+            Pattern.MULTILINE or Pattern.CASE_INSENSITIVE)
+
+    /**
+     * Render a markdown text message to Spannable.
+     *
+     * @param text The text message containing markdown syntax.
+     * @param quote An optional message to be quoted either by a quote or reply action.
+     * @param urls A list of urls to convert to markdown link syntax.
+     *
+     * @return A Spannable with the parsed markdown.
+     */
+    fun renderMarkdown(text: String, quote: MessageViewModel?, urls: List<Url>): CharSequence {
```

Comment:
Lets make it suspend and run as `launch(CoomonPool)`

Reply:
It's not needed (on this branch). Since `renderMarkdown()` is only called at `MessageViewModel` `init {}` which in turn is only invoked on a suspend function:

https://github.com/RocketChat/Rocket.Chat.Android/blob/b2d415bbadaaca7627802383b20eaa073c45d322/app/src/main/java/chat/rocket/android/chatroom/viewmodel/MessageViewModelMapper.kt#L18-L18

Reply:
But it's a good point. I'll revisit this later since I think I call `renderMarkdown()` at other places.

Reply:
But this suspend function is called from the UI context, so it will block UI anyway...

Reply:
But let's fix it later.

Reply:
Ok, saw now that it is called from a `CommonPool` context now... But coroutines are so cheap that there is no problem from calling one from another...

Reply:
No. I fixed it already. It's using `CommonPool`. See the line above.

Reply:
@luciofm I set `MessageViewModel` class as a data class but it's everything but not one. I didn't remove the `data` modifier from it solely for the free toString() for debugging. Are you working or are going to work on moving everything to the mapper?

Reply:
@leonardoaramaki yes, the intention is to have the Models just as simple POKOs (Plain Old Kotlin Objects), and move all the work to the mapper and maybe factories if the mapper becomes too complex.